README
Resurrection and enhancements Dug's all-time-classic network sniffer;

* Hidden command line options
* SNI and SSH decoder
* HTTP parsing improvements.
* Report every result only once (to fend of fuzzers and brute forcers).

Compare both branches: [Diff](https://github.com/hackerschoice/dsniff/compare/orig...main)

```sh
export ENV_ARGS="-i eth0 -P -v"
./dsniff
```

---

dsniff-2.3
----------

i wrote these tools with honest intentions - to audit my own network,
and to demonstrate the insecurity of cleartext / weakly-encrypted
network protocols and ad-hoc PKI. please do not abuse this software.

these programs require:

      Berkeley DB - http://www.sleepycat.com/
      OpenSSL - http://www.openssl.org/
      libpcap - http://www.tcpdump.org/
      libnids - http://www.packetfactory.net/Projects/Libnids/
      libnet - http://www.packetfactory.net/Projects/Libnet/

built and tested on OpenBSD, Linux, and Solaris. YMMV.

what's here:

arpspoof
	redirect packets from a target host (or all hosts) on the LAN
	intended for another local host by forging ARP replies. this
	is an extremely effective way of sniffing traffic on a switch.
	kernel IP forwarding (or a userland program which accomplishes
	the same, e.g. fragrouter :-) must be turned on ahead of time.

dnsspoof
	forge replies to arbitrary DNS address / pointer queries on
	the LAN. this is useful in bypassing hostname-based access
	controls, or in implementing a variety of man-in-the-middle
	attacks (HTTP, HTTPS, SSH, Kerberos, etc).

dsniff
	password sniffer. handles FTP, Telnet, SMTP, HTTP, POP,
	poppass, NNTP, IMAP, SNMP, LDAP, Rlogin, RIP, OSPF, PPTP
	MS-CHAP, NFS, VRRP, YP/NIS, SOCKS, X11, CVS, IRC, AIM, ICQ,
	Napster, PostgreSQL, Meeting Maker, Citrix ICA, Symantec
	pcAnywhere, NAI Sniffer, Microsoft SMB, Oracle SQL*Net, Sybase
	and Microsoft SQL auth info.

	dsniff automatically detects and minimally parses each
	application protocol, only saving the interesting bits, and
	uses Berkeley DB as its output file format, only logging
	unique authentication attempts. full TCP/IP reassembly is
	provided by libnids(3) (likewise for the following tools as
	well).

filesnarf
	saves selected files sniffed from NFS traffic in the current
	working directory.

macof
	flood the local network with random MAC addresses (causing
	some switches to fail open in repeating mode, facilitating
	sniffing). a straight C port of the original Perl Net::RawIP
	macof program.

mailsnarf
	a fast and easy way to violate the Electronic Communications
	Privacy Act of 1986 (18 USC 2701-2711), be careful. outputs
	selected messages sniffed from SMTP and POP traffic in Berkeley
	mbox format, suitable for offline browsing with your favorite
	mail reader (mail -f, pine, etc.).

msgsnarf
	record selected messages from sniffed AOL Instant Messenger,
	ICQ 2000, IRC, and Yahoo! Messenger chat sessions.

sshmitm
	SSH monkey-in-the-middle. proxies and sniffs SSH traffic
	redirected by dnsspoof(8), capturing SSH password logins, and
	optionally hijacking interactive sessions. only SSH protocol
	version 1 is (or ever will be) supported - this program is far
	too evil already.

sshow
	SSH traffic analysis tool. analyzes encrypted SSH-1 and SSH-2
	traffic, identifying authentication attempts, the lengths of
	passwords entered in interactive sessions, and command line
	lengths.

tcpkill
	kills specified in-progress TCP connections (useful for
	libnids-based applications which require a full TCP 3-whs for
	TCB creation).

tcpnice
	slow down specified TCP connections via "active" traffic
	shaping. forges tiny TCP window advertisements, and optionally
	ICMP source quench replies.

urlsnarf
	output selected URLs sniffed from HTTP traffic in CLF
	(Common Log Format, used by almost all web servers), suitable
	for offline post-processing with your favorite web log
	analysis tool (analog, wwwstat, etc.).

webmitm
	HTTP / HTTPS monkey-in-the-middle. transparently proxies and
	sniffs web traffic redirected by dnsspoof(8), capturing most
	"secure" SSL-encrypted webmail logins and form submissions.

webspy
	sends URLs sniffed from a client to your local Netscape
	browser for display, updated in real-time (as the target
	surfs, your browser surfs along with them, automagically).
	a fun party trick. :-)

-d.

---
http://www.monkey.org/~dugsong/


debian/changelog
dsniff (2.4b1+debian-31) unstable; urgency=medium

  [ Marcos Fouces ]
  * Acknowledge Janitor commits.
  * Update d/copyright.

  [ Debian Janitor ]
  * Update standards version to 4.6.0, no changes needed.
  * Remove constraints unnecessary since buster (oldstable)
  * Update standards version to 4.6.1, no changes needed.

 -- Marcos Fouces <marcos@debian.org>  Fri, 04 Nov 2022 18:22:37 +0100

dsniff (2.4b1+debian-30) unstable; urgency=medium

  [ Samuel Henrique ]
  * Add salsa-ci.yml
  * Configure git-buildpackage for Debian

  [ Deb Janitor ]
  * Use secure URI in Homepage field.
  * Set debhelper-compat version in Build-Depends.
  * Fix day-of-week for changelog entries 2.4b1-10, 2.4b1-3.6, 2.4b1-3.5.

  [ Marcos Fouces ]
  * Bump to debhelper-compat 12.
  * Add Rules-Requires-Root: no in control file.
  * Bump to Standards-Version 4.5.0
  * Add modified 37_fix-lib-and-share-dirs.patch (Closes: #919631).
  * Remove -Wl,--as-needed linker flags as they are no more needed
    with Bullseye toolchain.
  * Update uploader email to @debian.org.
  * Add 38_fix-pcap_init.patch (Closes: #980588)
  * Bump to debhelper-compat level 13.
  * Bump to Standards-Version 4.5.1 (no changes required).
  * Build against libtirpc as a separate library.

 -- Marcos Fouces <marcos@debian.org>  Wed, 10 Feb 2021 18:26:44 +0100

dsniff (2.4b1+debian-29) unstable; urgency=medium

  [ Raphaël Hertzog ]
  * Update team maintainer address to
    Debian Security Tools <team+pkg-security@tracker.debian.org>
  * Update Vcs-Git and Vcs-Browser for the move to salsa.debian.org

  [ Samuel Henrique ]
  * d/copyright: Remove trailing whitespace and align author names

  [ Marcos Fouces ]
  * Update to dh 11
  * Remove trailing whitespaces.
  * Fix spelling errors spotted by Lintian.
  * Update copyright years.
  * Add proper Files-excluded field in d/copyright.
  * Bump to Standards-revision 4.2.1 (no changes required).
  * Use secure URL in d/watch file.
  * Bump to uscan version 4.

 -- Marcos Fouces <marcos.fouces@gmail.com>  Sun, 04 Nov 2018 22:22:12 -0200

dsniff (2.4b1+debian-28) unstable; urgency=medium

  [ Lukas Schwaighofer ]
  * fix implicit declarations compiler warning
    - switch to C99 uint64_t, remove now unnecessary xdr_u_int64_t function
    - include missing string.h for memset
  * bump Standards version to 4.0.0 (no change required)

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Fri, 21 Jul 2017 20:05:34 +0200

dsniff (2.4b1+debian-27) unstable; urgency=high

  * Fix a parallel FTBFS due to missing Makefile dependencies.  Thanks to
    Adrian Bunk (Closes: #869086).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Thu, 20 Jul 2017 20:00:38 +0200

dsniff (2.4b1+debian-26) unstable; urgency=medium

  * Team upload.

  [ Marcos Fouces ]
  * Fix 27_arpspoof-allow-use-of-multiple-interface.patch
    Thanks to João Salavisa. Closes: #706766
  * Add 34_Add_CPPFLAGS.patch to build binaries with
    all default Debian compiler flags.

  [ Lukas Schwaighofer ]
  * no longer export CC, cross compiling works fine without on dh 10
  * avoid unnecessary linking by passing -Wl,--as-needed
  * pass prefix= to dh_auto_install instead of calling make directly
  * remove outdated information from README.Debian, document notable changes
    from upstream instead

  [ Raphaël Hertzog ]
  * Drop Recommends on libx11-6, it's already in Depends via
    shlibs:Depends.

 -- Raphaël Hertzog <hertzog@debian.org>  Thu, 15 Jun 2017 18:20:50 +0200

dsniff (2.4b1+debian-25) unstable; urgency=medium

  * debian/control: Added myself to Uploaders.
  * Make sure libmissing.a is built before PROGS (Closes: #860611).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Wed, 19 Apr 2017 20:15:27 +0200

dsniff (2.4b1+debian-24) unstable; urgency=medium

  * Fix FTCBFS: Pass triplet-prefixed CC to configure.
    Thanks to Helmut Grohne (Closes: #852360).
  * Add four patches from Fedora:
    (Closes: #715646, #716355, #716457, #716458)
    + 29_libnet_name2addr4.patch
    + 30_pntohl_shift.patch
    + 31_sysconf_clocks.patch
    + 32_rpc_segfault.patch (Closes: #855869)
  * Polish, reorder and refresh patches.
  * Add 33_sshcrypto_DES.patch
  * Improve dep3 description in 33_sshcrypto_DES.patch
  (as request by Stretch release team)

 -- Marcos Fouces <mfouces@yahoo.es>  Mon, 13 Mar 2017 18:34:19 +0100

dsniff (2.4b1+debian-23) unstable; urgency=medium

  * Assign to pkg-security team (Closes: #847505)
  * Add 24_Fix-OpenSSL1.1.0-Build.patch.
    Thanks to Christoph Biedl. Closes (#828287)
  * Bump to dh compat level 10
  * Update copyright file
  * Bump to standards-revision 3.9.8
  * Fix copyright file
  * Add hardening options
  * Add 25_fix-spelling-errors.patch

 -- Marcos Fouces <mfouces@yahoo.es>  Tue, 06 Dec 2016 12:08:15 +0100

dsniff (2.4b1+debian-22.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix TDS decoder patch (Closes: #712648).

 -- Andrew Shadura <andrewsh@debian.org>  Sat, 30 Nov 2013 21:29:23 +0100

dsniff (2.4b1+debian-22) unstable; urgency=low

  * Acknowledge NMU.
     + Thanks Hilko Bengen
  * Bump standards to 3.9.3.
  * Added a patches:
     + Fix POP decoder in dsniff (Closes: #647583).
     + Add both communication partners in arpspoof (Closes: #650749).
     + Allow multiple targets to be imitated simultaniously (Closes: #650751).
     + arpspoof: allow the selection of source hw address (Closes: #650752).
     + Thanks Stefan Tomanek.

 -- William Vera <billy@billy.com.mx>  Tue, 06 Mar 2012 22:42:43 -0600

dsniff (2.4b1+debian-21.1) unstable; urgency=low

  * Non-maintainer upload.
  * Updated urlsnarf timestamp patch (Closes: #636202)

 -- Hilko Bengen <bengen@debian.org>  Tue, 06 Sep 2011 10:15:26 +0200

dsniff (2.4b1+debian-21) unstable; urgency=low

  * Changed patch system to quilt from dpatch.
  * Updated debian/README.source for quilt.
  * Removed lintian override.
  * Cleanup the debian/rules.
  * Build-Depend changed to libdb-dev (>= 4.7) (Closes: #548478).

sshow.c
/*
 * sshow.c
 *
 * Passive SSH traffic analysis.
 *
 * http://www.openwall.com/advisories/OW-003-ssh-traffic-analysis.txt
 *
 * Copyright (c) 2000-2001 Solar Designer <solar@openwall.com>
 * Copyright (c) 2000 Dug Song <dugsong@monkey.org>
 *
 * $Id: sshow.c,v 1.2 2001/03/19 06:52:15 dugsong Exp $
 */

#include "config.h"

#include <sys/types.h>
#include <sys/times.h>
#include <time.h>

#include <netinet/in_systm.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <err.h>
#include <nids.h>
#include <pcap.h>

#include "pcaputil.h"

#if !defined(NIDS_MAJOR) || (NIDS_MAJOR == 1 && NIDS_MINOR < 15)
#error This program requires libnids 1.15+
#endif

#define HISTORY_SIZE			16

typedef struct {
	u_int min, max;
} range;

typedef struct {
	int direction;			/* 0 for client to server */
	clock_t timestamp;		/* timestamp of this packet */
	u_int cipher_size;		/* ciphertext size */
	range plain_range;		/* possible plaintext sizes */
} record;

struct history {
	record packets[HISTORY_SIZE];	/* recent packets (circular list) */
	int index;			/* next (free) index into packets[] */
	u_int directions;		/* recent directions (bitmask) */
	clock_t timestamps[2];		/* last timestamps in each direction */
};

struct line {
	int input_count;		/* input packets (client to server) */
	int input_size;			/* input size (estimated) */
	int input_last;			/* last input packet size */
	int echo_count;			/* echo packets (server to client) */
};

struct session {
	int protocol;		/* -1 not SSH, 0 unknown, 1 or 2 once known */
	int state;		/* 1 after username, 2 after authentication */
	int compressed;		/* whether compression is known to be used */
	struct history history;	/* session history */
	struct line line;	/* current command line */
};

static int debug = 0;

static clock_t now;

static void
usage(void)
{
	fprintf(stderr, "Usage: sshow [-d] [-i interface | -p pcapfile]\n");
	exit(1);
}

static clock_t
add_history(struct session *session, int direction,
	    u_int cipher_size, range *plain_range)
{
	record *current;
	clock_t delay;

	current = &session->history.packets[session->history.index++];
	session->history.index %= HISTORY_SIZE;

	current->direction = direction;
	current->timestamp = now;
	current->cipher_size = cipher_size;
	current->plain_range = *plain_range;

	session->history.directions <<= 1;
	session->history.directions |= direction;

	delay = now - session->history.timestamps[direction];
	session->history.timestamps[direction] = now;

	return (delay);
}

static record *
get_history(struct session *session, int age)
{
	int index;

	index = session->history.index + (HISTORY_SIZE - 1) - age;
	index %= HISTORY_SIZE;

	return (&session->history.packets[index]);
}

static char *
s_saddr(struct tcp_stream *ts)
{
	static char output[32];

	snprintf(output, sizeof(output), "%s:%u",
		inet_ntoa(*((struct in_addr *)&ts->addr.saddr)),
		ts->addr.source);
	return (output);
}

static char *
s_daddr(struct tcp_stream *ts)
{
	static char output[32];

	snprintf(output, sizeof(output), "%s:%u",
		inet_ntoa(*((struct in_addr *)&ts->addr.daddr)),
		ts->addr.dest);
	return (output);
}

static char *
s_range(range *range)
{
	static char output[32];

	snprintf(output, sizeof(output),
		range->min == range->max ? "%u" : "%u to %u",
		range->min, range->max);
	return (output);
}

static void
print_data(struct half_stream *stream, u_int count)
{
	u_int i;
	int printable;

	printable = 1;
	for (i = 0; i < count; i++) {
		printf("%02x%c", (int)(u_char)stream->data[i],
			i < count - 1 && i % 24 != 23
			? ' ' : '\n');
		printable &=
			isprint(stream->data[i]) ||
			stream->data[i] == '\n';
	}
	if (printable && count >= 4 && !memcmp(stream->data, "SSH-", 4))
		fwrite(stream->data, count, 1, stdout);
}

static u_int
ssh1_plain_size(struct half_stream *stream)
{
	if (stream->count_new < 4) return (0);

	return (u_int)(u_char)stream->data[3] |
		((u_int)(u_char)stream->data[2] << 8) |
		((u_int)(u_char)stream->data[1] << 16) |
		((u_int)(u_char)stream->data[0] << 24);
}

static u_int
ssh1_cipher_size(struct half_stream *stream)
{
	return (4 + ((ssh1_plain_size(stream) + 8) & ~7));
}

static range *
ssh1_plain_range(struct half_stream *stream)
{
	static range output;

	output.min = output.max = ssh1_plain_size(stream) - 5;
	return (&output);
}

static range *
ssh2_plain_range(struct half_stream *stream)
{
	static range output;

	output.max = stream->count_new - 16;
	/* Assume min padding + 8-byte cipher blocksize */
	output.min = output.max - 7;
	if ((int)output.min < 0) output.min = 0;
	return (&output);
}

static void
client_to_server(struct tcp_stream *ts, struct session *session,
		 u_int cipher_size, range *plain_range)
{
	clock_t delay;
	int payload;
#if defined(_SC_CLK_TCK)
	long CLK_TCK = sysconf(_SC_CLK_TCK);
#endif

	delay = add_history(session, 0, cipher_size, plain_range);

	if (debug)
		printf("- %s -> %s: DATA (%s bytes, %.2f seconds)\n",
			s_saddr(ts), s_daddr(ts), s_range(plain_range),
			(float)delay / CLOCKS_PER_SEC);
	if (debug > 1)
		print_data(&ts->server, cipher_size);

	payload = plain_range->min;
	if (session->state == 2 && payload > 0) {
		session->line.input_count++;
		session->line.input_last = payload;
		if (session->protocol == 1)
			payload -= 4;
		else {
			payload -= 20 + 1;
			/* Assume several SSH-2 packets in this IP packet */
			if (payload % 44 == 0) {
				session->line.input_count += payload / 44;
				/* One character per SSH-2 packet (typical) */
				payload += payload / 44;
				payload %= 44;
			}
			payload++;
		}
		if (payload <= 0) {
			if (payload < 0 && !session->compressed &&
			    session->protocol == 1) {
				session->compressed = 1;
				printf("+ %s -> %s: Compression detected, "
					"guesses will be much less reliable\n",
					s_saddr(ts), s_daddr(ts));
			}
			payload = 1;
		}
		session->line.input_size += payload;
	}
}

static void
server_to_client(struct tcp_stream *ts, struct session *session,
		 u_int cipher_size, range *plain_range)
{
	clock_t delay;
	int skip;
	range string_range;
#if defined(_SC_CLK_TCK)
	long CLK_TCK = sysconf(_SC_CLK_TCK);
#endif
	
	delay = add_history(session, 1, cipher_size, plain_range);
	
	if (debug)
		printf("- %s <- %s: DATA (%s bytes, %.2f seconds)\n",
		       s_saddr(ts), s_daddr(ts), s_range(plain_range),
		       (float)delay / CLOCKS_PER_SEC);
	if (debug > 1)
		print_data(&ts->client, cipher_size);
	
/*
 * Some of the checks may want to skip over multiple server responses.
 * For example, there's a debugging packet sent for every option found
 * in authorized_keys, but we can't use those packets in our pattern.
 */
	skip = 0;
	while (((session->history.directions >> skip) & 3) == 3)
		if (++skip > HISTORY_SIZE - 5) break;
	
	if (session->state == 0 &&
	    session->protocol == 1 &&
	    ((session->history.directions >> skip) & 7) == 5 &&
	    plain_range->min == 0 &&
	    get_history(session, skip + 1)->plain_range.min > 4 &&
	    get_history(session, skip + 2)->plain_range.min == 0) {
		session->state = 1;
		string_range = get_history(session, skip + 1)->plain_range;
		string_range.min -= 4; string_range.max -= 4;
		printf("+ %s -> %s: GUESS: Username length is %s\n",
		       s_saddr(ts), s_daddr(ts), s_range(&string_range));
		return;
	}
	
	if (session->state == 1 &&
#ifdef USE_TIMING
	    now - get_history(session, 2)->timestamp >= CLOCKS_PER_SEC &&
#endif
	    session->protocol == 1 &&
	    (session->history.directions & 7) == 5 &&
	    plain_range->min == 0 &&
	    get_history(session, 1)->plain_range.min > 4 &&
	    get_history(session, 2)->plain_range.min == 0) {
		session->state = 2;
		string_range = get_history(session, 1)->plain_range;
		string_range.min -= 4; string_range.max -= 4;
		printf("+ %s -> %s: GUESS: Password authentication, "
		       "password length %s %s%s\n",
		       s_saddr(ts), s_daddr(ts),
		       string_range.min == 32 ? "appears to be" : "is",
		       s_range(&string_range),
		       string_range.min == 32 ? " (padded?)" : "");
	}
	
	if (session->state == 0 &&
	    session->protocol == 2 &&
	    (session->history.directions & 7) == 5) {
		if (plain_range->min == 4 + 9) {
			string_range = get_history(session, 1)->plain_range;
			
			if (string_range.min > 500 && string_range.min < 600) {
				session->state = 2;
				printf("+ %s -> %s: GUESS: DSA "
				       "authentication accepted\n",
				       s_saddr(ts), s_daddr(ts));
			} else
				if (string_range.min > 42 + 9) {
					session->state = 2;
					printf("+ %s -> %s: GUESS: Password "
					       "authentication accepted\n",
					       s_saddr(ts), s_daddr(ts));
				}
		} else if (plain_range->min > 12 + 9 &&
			   plain_range->min < 56 + 9) {
			string_range = get_history(session, 1)->plain_range;
			
			if (string_range.min > 500 && string_range.min < 600)
				printf("+ %s -> %s: GUESS: DSA "
				       "authentication failed\n",
				       s_saddr(ts), s_daddr(ts));
			else if (string_range.min > 42 + 9)
				printf("+ %s -> %s: GUESS: Password "
				       "authentication failed\n",
				       s_saddr(ts), s_daddr(ts));
		}
	}
	
	if (session->state == 1 &&
	    session->protocol == 1 &&
	    (session->history.directions & 3) == 1 &&
	    plain_range->min == 0 &&
	    get_history(session, 1)->plain_range.min == 130) {
		printf("+ %s -> %s: GUESS: RSA authentication refused\n",
		       s_saddr(ts), s_daddr(ts));
	}
	
	if (session->state == 1 &&
	    session->protocol == 1 &&
	    skip >= 1 &&
	    ((session->history.directions >> (skip - 1)) & 037) == 013 &&
	    plain_range->min == 0 &&
	    get_history(session, skip - 1 + 2)->plain_range.min == 16 &&
	    get_history(session, skip - 1 + 3)->plain_range.min == 130 &&
	    get_history(session, skip - 1 + 4)->plain_range.min == 130) {
		char *what;
		
		switch (get_history(session, 1)->plain_range.min - 4) {
		case 28:
			/* "RSA authentication accepted." */
			session->state = 2;
			if (skip > 1 && (what = alloca(64))) {
				snprintf(what, 64, "accepted "
					 "(%d+ authorized_keys option%s)",
					 skip - 1, skip - 1 == 1 ? "" : "s");
				break;
			}
			what = "accepted";
			break;
			
		case 47:
			/* "Wrong response to RSA authentication challenge." */
			what = "failed";
			break;
			
		default:
			what = "???";
		}
		printf("+ %s -> %s: GUESS: RSA authentication %s\n",
		       s_saddr(ts), s_daddr(ts), what);
	}
	
	if (session->state == 2) {
		session->line.echo_count++;
		
		/* Check for backspace */
		if (session->protocol == 1 && !session->compressed &&
		    plain_range->min == 4 + 3 &&
		    session->line.input_size >= 2)
			session->line.input_size -= 2;
		
		if (plain_range->min > 4 + session->line.input_last &&
		    session->line.input_count >= 2 &&
		    session->line.input_size >= 2) {
			int size;
			char *what;
			
			size = session->line.input_size;
			if (session->line.echo_count + 1 >=
			    session->line.input_count &&
			    size <= (session->line.input_count << 2) &&
			    size < 0x100) {
				what = "(command) line";
			}
			else {
				if (session->line.echo_count <= 2 &&
				    size <= (session->line.input_count << 1) &&
				    size >= 2 + 1 && size <= 40 + 1) {
					what = "password";
				}
				else what = NULL;
			}
			if (debug) {
				printf("- %s -> %s: sent %d packets "
				       "(%d characters), seen %d replies\n",
				       s_saddr(ts), s_daddr(ts),
				       session->line.input_count, size,
				       session->line.echo_count);
			}
			if (what) {
				printf("+ %s -> %s: GUESS: "
				       "a %s of %d character%s\n",
				       s_saddr(ts), s_daddr(ts),
				       what, size - 1, size == 2 ? "" : "s");
			}
		}
		if (plain_range->min <= 0 ||
		    plain_range->min > 4 + session->line.input_last) {
			session->line.input_count = 0;
			session->line.input_size = 0;
			session->line.echo_count = 0;
		}
	}
}

static void
process_data(struct tcp_stream *ts, struct session *session)
{
	u_int have, need;
	char *lf;
	
	if (session->protocol < 0) return;
	
	if (ts->client.count_new &&
	    (have = ts->client.count - ts->client.offset)) {
		switch (session->protocol) {
		case 1:
			if (have < (need = ssh1_cipher_size(&ts->client))) {
				if (debug) {
					printf("- %s <- %s: got %u of "
					       "%u bytes\n", s_saddr(ts),
					       s_daddr(ts), have, need);
				}
				nids_discard(ts, 0);
				return;
			}
			if (have != need && debug) {
				printf("- %s <- %s: left %u bytes\n",
				       s_saddr(ts), s_daddr(ts),
				       have - need);
			}
			nids_discard(ts, need);
			server_to_client(ts, session, need,
				ssh1_plain_range(&ts->client));
			return;
			
		case 2:
			server_to_client(ts, session, have,
					 ssh2_plain_range(&ts->client));
			return;
			
		default:
			break;
		}
	}
	if (ts->server.count_new &&
	    (have = ts->server.count - ts->server.offset)) {
		if (!session->protocol) {
			lf = (char *)memchr(ts->server.data, '\n', have);
			if (have < 7 || (!lf && have < 0x100)) {
				nids_discard(ts, 0);
				return;
			}
			if (lf && !memcmp(ts->server.data, "SSH-", 4))
				session->protocol = ts->server.data[4] - '0';
			/* some clients announce SSH-1.99 instead of SSH-2.0 */
			if (session->protocol == 1 &&
			    ts->server.data[5] == '.' &&
			    ts->server.data[6] == '9') {
				session->protocol = 2;
			}
			if (session->protocol != 1 && session->protocol != 2) {
				session->protocol = -1;
				if (debug) {
					printf("- %s -> %s: not SSH\n",
					       s_saddr(ts), s_daddr(ts));
				}
				return;
			}
			need = lf - ts->server.data + 1;
			nids_discard(ts, need);
			printf("+ %s -> %s: SSH protocol %d\n",
			       s_saddr(ts), s_daddr(ts), session->protocol);
			if (debug)
				print_data(&ts->server, have);
			return;
		}
		
		switch (session->protocol) {
		case 1:
			if (have < (need = ssh1_cipher_size(&ts->server))) {
				if (debug) {
					printf("- %s -> %s: got %u of "
					       "%u bytes\n", s_saddr(ts),
					       s_daddr(ts), have, need);
				}
				nids_discard(ts, 0);
				return;
			}
			if (have != need && debug) {
				printf("- %s -> %s: left %u bytes\n",
				       s_saddr(ts), s_daddr(ts),
				       have - need);
			}
			nids_discard(ts, need);
			client_to_server(ts, session, need,
					 ssh1_plain_range(&ts->server));
			return;
			
		case 2:
			client_to_server(ts, session, have,
					 ssh2_plain_range(&ts->server));
		}
	}
}

static void
process_event(struct tcp_stream *ts, struct session **session)
{
	struct tms buf;
	char *what;
	
	now = times(&buf);
	what = NULL;
	
	switch (ts->nids_state) {
	case NIDS_JUST_EST:
		ts->client.collect = 1;
		ts->server.collect = 1;
		if (debug) {
			printf("- %s -> %s: ESTABLISHED\n",
			       s_saddr(ts), s_daddr(ts));
		}
		if (!(*session = calloc(1, sizeof(**session)))) {
			err(1, "calloc");
		}
		(*session)->history.timestamps[0] = now;
		(*session)->history.timestamps[1] = now;
		return;
		
	case NIDS_CLOSE:
		what = "CLOSED";
		
	case NIDS_RESET:
		if (!what) what = "RESET";
		
	case NIDS_TIMED_OUT:
		if (!what) what = "TIMED OUT";
		if ((*session)->protocol > 0) {
			printf("+ %s -- %s: %s\n",
			       s_saddr(ts), s_daddr(ts), what);
		}
		else if (debug) {
			printf("- %s -- %s: %s\n",
			       s_saddr(ts), s_daddr(ts), what);
		}
		free(*session);
		return;
		
	case NIDS_DATA:
		process_data(ts, *session);
		return;
	}
}

static void
null_syslog(int type, int errnum, struct ip *iph, void *data)
{
}

static void
cleanup(int signum)
{
	exit(0);	/* Just so that atexit(3) jobs are called */
}

int
main(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;
	int c;
	
	while ((c = getopt(argc, argv, "di:p:h?")) != -1) {
		switch (c) {
		case 'd':
			debug++;
			break;
		case 'i':
			nids_params.device = optarg;
			break;
		case 'p':
			nids_params.filename = optarg;
			break;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;
	
	signal(SIGTERM, cleanup);
	signal(SIGINT, cleanup);
	signal(SIGHUP, cleanup);
	
	setlinebuf(stdout);
	
	if (argc > 0) {
		nids_params.pcap_filter = copy_argv(argv);
	}
	else nids_params.pcap_filter = "tcp";
	
	nids_params.syslog = null_syslog;
	nids_params.scan_num_hosts = 0;
	nids_params.one_loop_less = 1;
	
	if (!nids_init())
		errx(1, "nids_init: %s", nids_errbuf);
	
	nids_register_tcp(process_event);

        if (nids_params.pcap_filter != NULL) {
                if (nids_params.filename == NULL) {
                        warnx("listening on %s [%s]", nids_params.device,
                              nids_params.pcap_filter);
                }
                else {
                        warnx("using %s [%s]", nids_params.filename,
                              nids_params.pcap_filter);
                }
        }
        else {
                if (nids_params.filename == NULL) {
                    warnx("listening on %s", nids_params.device);
                }
                else {
                    warnx("using %s", nids_params.filename);
                }
        }

	nids_run();
	
	return (0);
}
